src/Data/BTree/Primitives.hs

```diff
@@ -61,6 +89,16 @@ data Tree key val where
 deriving instance (Show key, Show val) => Show (Node height key val)
 deriving instance (Show key, Show val) => Show (Tree key val)
 
+instance (Binary key, Binary val) => Binary (Tree key val) where
+    put (Tree height rootId) = put height >> put rootId
+    get = Tree <$> get <*> get
+
+instance (Typeable key, Typeable val, Eq key, Eq val) => Eq (Tree key val) where
```

Comment:
Here you are really comparing `Tree`s for same shape. This is stronger than the usual "are equivalent finite maps". This is only used in testing right? Then I suggest not reusing the Eq class but instead implement normal function (or another class) and move this to an `Internal` module.